languagetool-core/src/main/java/org/languagetool/rules/patterns/Substrings.java

```diff
@@ -0,0 +1,187 @@
+/* LanguageTool, a natural language style checker
+ * Copyright (C) 2020 Peter Gromov
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+package org.languagetool.rules.patterns;
+
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+
+/**
+ * A list of strings that are expected to occur in a larger string in the given order,
+ * with optional requirements for them to start/end the text.
+ * This class is used to avoid more expensive regular expression matching where possible.
+ */
+class Substrings {
```

Comment:
This class would be a good candidate for a unit test, wouldn't it?

Reply:
It's more of an implementation detail of `StringMatcher`, extracted to keep it smaller. So most of it is already tested (indirectly) in `StringMatcherTest`.